### PR TITLE
Introduce Plug-in System

### PR DESCRIPTION
**Pull request by benoitf**
This PR brings a new way to extend Theia by using plugins (in addition to extension)

The plugin model is having (compared to Extension) :
- code runtime isolation (separated node instance or in a web worker)
- self-contained (no need to download external dependencies)
- no dependency requirement (like inversify) except the dependency to the API type. `@theia/plugin`
- no need to rebuild the IDE
- simple API with a single entrypoint (`theia` object)
- lifecycle with a start and stop method

More details on the original issue: https://github.com/theia-ide/theia/issues/1482

[Documentation on the API is provided](https://github.com/theia-demo-plugins/wiptheia/blob/master/packages/plugin/API.md)

There are still missing API but it will come as enhancements to the current model.

Here is a quick video on youtube (click on the image)
[Plugin in a running Che](https://www.youtube.com/watch?v=2DbBo1WDGHE)

Please feel to add any reviewer to review the PR.

### CODE REVIEW

**Comment by akosyakov** [comment]
Impressive work!

What are reasons/benefits of:
- changing the travis file;
- deviating from the code organization rules: https://github.com/theia-ide/theia/wiki/Code-Organization
- using the custom rcp-protocol instead of the core messaging module

**Comment by benoitf** [comment]
@akosyakov 
1.
on the travis file, it was to publish it on a different scope to try it. But this commit is flagged as `TO BE DELETED` before the merge. So of course, we won't merge any travis change at all (or appveyor or any other)


2. On code organization rule, it was to enforce a clear separation on what will be used in "forked" instance to the remaining one (by default on the linked document, the code is either on browser or or node but never in a `forked` instance

3.
On custom RCP-Protocol, you mean by using vscode-jsonrpc one ? (I tried to search because I was almost sure we already discussed that point but I didn't find the content online) But if I remember well we said we could postpone that RPC rewrite a bit later.

**Comment by akosyakov** [comment]
@benoitf ok, could you adjust commits and i will have a look once more?

**Comment by benoitf** [comment]
@akosyakov you think we can reach merging state soon ? (if I adjust commits to get rid of .Travis CI changes)

**Comment by akosyakov** [comment]
@benoitf if it does not break any existing code i am fine to merge it that you can continue on master. The package structured, rewriting RPC and any other issues can be improved with follow-up PRs.

File: packages/plugin-ext/src/common/editor-options.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+/**
+ * The style in which the editor's cursor should be rendered.
+ */
+export enum TextEditorCursorStyle {
```
Comment (AlexTugarev) [inline packages/plugin-ext/src/common/editor-options.ts]: This looks like it's copied from vscode. If so, please attribute with MIT license in the header.

**Comment by benoitf** [comment]
@akosyakov ok rebased and removing travis ci commits

**Comment by akosyakov** [comment]
@benoitf it seems that not all commits are signed off

**Comment by benoitf** [comment]
@akosyakov now DCO Sign-off is ok
appveyor build is failing due to the API rate limit
```
Downloading ripgrep failed: Error: API rate limit exceeded for 67.225.139.144. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**Comment by benoitf** [comment]
@akosyakov now, is it ok for you as well ?

**Pull request merged**